[Bibi] 미션2 - 체스판 생성 PR입니다.

bibi6666667 opened the pull request:
안녕하세요. 백엔드반 Bibi입니다.😊
미션 2단계 Pull 요청 드립니다.

<작업 내용>
- (기본 생성자)자바 생성자를 공부하며 Pawn클래스에 매개변수 없는 생성자를 만들고, 컴파일 에러를 해결했습니다.
- (상수)자바 상수에 대해 공부하고, Pawn 클래스에 상수를 만들어 중복코드를 제거했습니다.
- (체스판)Board클래스를 만들고 BoardTest를 추가해 제시된 테스트를 완료했습니다. 또, BoardTest에서 Pawn이외 객체 추가시 컴파일 에러가 나는 것을 확인했습니다.
- (패키지 분리)기존 패키지 bibi를 삭제하고 pieces, chess패키지를 만들어 각각 Pawn과 Board를 넣었습니다.
- (중복 제거) 각 클래스의 중복 코드를 제거했습니다. 인스턴스변수를 사용하고, 메소드를 분리하는 방법으로 중복을 제거했습니다. 테스트 클래스의 경우 @BeforeEach 사용법을 배우고 적용해 중복을 제거했습니다.

<학습 내용>
기본 생성자, 상수, 패키지 분리하는 법과 import 복습, JUnit과 AssertJ의 개념 및 기초적인 사용법(assert메소드, 어노테이션)

부족한 점이 많은 코드라고 생각합니다. 어떤 것이든 괜찮으니 많은 조언과 피드백 부탁드리겠습니다! 
매 미션마다 코드리뷰 신경 써 주시는 점 감사드립니다🙇‍♀️🙇‍♀️ 좋은 저녁 되세요!

bibi6666667 committed:
[step1] CONFLICT해결) rebase 중 충돌 부분 수정

rebase하기 전 수정했던 .gitignore파일에서 충돌이 있어서
해당 부분을 수정 후, 다시 rebase했습니다.

bibi6666667 committed:
[step2] 추가) 색이 없는 Pawn 만들기 (매개변수 없는 생성자)

PawnTest - 색이 없는 Pawn을 생성하는 테스트를 추가.

컴파일 에러 발생 원인 : 매개변수 없는 기본 생성자는
원래 자동으로 생성되지만, 다른 생성자를 만든 경우
기본 생성자는 자동으로 생성되지 않는다.
따라서 생성자를 여러 개 사용하는 경우
기본 생성자도 별도로 선언해 주어야 한다.

컴파일 에러 해결 : Pawn 클래스에 기본 생성자를 선언한다.

Pawn - 색이 없는 Pawn을 기본 생성자로 생성한 경우,
흰색 말을 생성하도록 두 번쨰 생성자를 추가.

bibi6666667 committed:
[step1] 수정) Pawn클래스의 final 삭제 및 멤버변수 color에final 추가

(호눅스의 피드백 반영)
클래스에 final을 사용한 의도는 Pawn클래스, 특히 color변수를
외부에서 변경하지 못하게 하는 것이었습니다.
오늘 그룹리뷰에서 질문해 보니, 그런 의도라면 final을
클래스보다는 color변수에 선언하는 것이 낫다고들 하셔서
해당하는 부분을 수정했습니다.

bibi6666667 committed:
[step2] 수정) Pawn클래스의 final 삭제 및 멤버변수 color에final 추가

(호눅스의 피드백 반영)
클래스에 final을 사용한 의도는 Pawn클래스, 특히 color변수를
외부에서 변경하지 못하게 하는 것이었습니다.
오늘 그룹리뷰에서 질문해 보니, 그런 의도라면 final을
클래스보다는 color변수에 선언하는 것이 낫다고들 하셔서
해당하는 부분을 수정했습니다.

bibi6666667 committed:
[step2] 생성) BLACK, WHITE 상수를 생성해 기존 색상을 대체함

Pawn과 PawnTest 클래스에서 중복 사용되는 "white"와 "black"을,
Pawn클래스에 상수로 선언해 사용했습니다.

bibi6666667 committed:
[step2] 생성) Board클래스 생성 및 메소드 구현

Board 클래스
- 체스판을 담당하는 Board클래스를 추가했습니다.

pawns
- 여러 개의 Pawn을 저장할 수 있게 ArrayList를 추가했습니다.

아래 메소드들은 BoardTest에서  필요했기 때문에 만들었습니다.

public void add(Pawn pawn)
: pawns에 Pawn을 추가합니다.
public int size()
: pawns에 존재하는 요소의 갯수를  리턴합니다.
public Pawn findPawn(int index
: pawns의 index번째 Pawn을 리턴합니다.

bibi6666667 committed:
[step2] 생성) BoardTest클래스와 테스트코드 생성

[step2] 2.BoardTest추가 에 따라 클래스와 테스트코드를 생성했습니다.

void putPawn()
: 흰색 및 검은색 Pawn을 체스판에 추가하고,
추가할 때마다 Pawn의 갯수와 추가된 Pawn을 확인하도록 했습니다.

bibi6666667 committed:
[step2] Merge branch 'step2' into bibi6666667

bibi6666667 committed:
[step2] 패키지분리, 접근제어자 수정)bibi->pieces, chess

src/main
기존 패키지 bibi를 chess로 변경했습니다.
새 패키지pieces를 생성하고, Pawn을 넣었습니다.
src/test
기존 패키지 bibi를 chess로 변경했습니다.
새 패키지 pieces를 생성하고, PawnTest를 넣었습니다.

패키지 수정 후 컴파일에러가 발생해 아래 조치를 취했습니다.

- Board, BoardTest, PawnTest에 import pieces.Pawn
- Board, BoardTest, Pawn, PawnTest의 package명시
- 접근제어자
: Pawn생성자들을 public으로 변경

bibi6666667 committed:
[step2] 추가) Pawn 외의 객체 추가시  에러 발생

BoardTest에서 폰 외의 객체를 보드에 추가하면 컴파일 에러가 발생하는지 확인했습니다.
(실제 에러가 발생하는 부분은 주석 처리했습니다)

bibi6666667 committed:
[step2] import static assertj, junit

PawnTest의 assertThat, assertEquals 메소드에 오류가 발생해
필요한 패키지를 import했습니다.

bibi6666667 committed:
[step2] 테스트클래스에 AssertJ메소드를 static import

중복제거를 하기 전에 PawnTest, BoartTest에 AssertJ 메소드를 정적 임포트했습니다.

bibi6666667 committed:
[step2] PawnTest의 create()메서드 기능 분리

create()가 흰색 폰과 검은색 폰을 둘 다 생성하고
잘 생성되었는지 테스트하는 역할이었는데,
한 메소드가 두 역할을 담당하고 있어서
기능 분리가 필요하다고 생각했습니다.

createWhitePawn()과 createBlackPawn()으로 나눈 뒤,
메소드에 맞는 내용으로 @DisplayName을 수정했습니다.

bibi6666667 committed:
[step2] 리팩토링) @BeforeEach로 중복 제거

BoardTest의 테스트케이스에서 중복되는 코드를 @BeforeEach를 사용해 묶어주었습니다.

bibi6666667 committed:
[step2] 리팩토링) putPawn() 기능 분리

putPawn() 메서드에서 흰색, 검은색 폰을 추가하고 확인하는
두 작업을 하고 있어서, 두 기능을  분리했습니다.

bibi6666667 committed:
Merge branch 'bibi6666667' of https://github.com/bibi6666667/java-chess into step2

bibi6666667 committed:
CONFLICT 해결

honux77 commented on src/main/java/bibi/Pawn.java:
```diff
@@ -1,13 +1,18 @@
+public class Pawn {
+    private final String color;
 
     Pawn(String color) {
         this.color = color;
     }
```
final 변수인데 생성자를 통해 값을 넣을 수 있다는 게 신기하지 않나요? 왜 그럴까요?

bibi6666667 replied:
final로 선언한 변수는 값을 변경할 수 없는 상수가 된다고 알고 있었습니다.
따라서, 위처럼 생성자에서 값을 할당하는 것은 값을 변경하는 게 아닌 초기화 과정이므로 문제가 없다. 까지 이해하고 있었습니다.
그런데 생각해 보니 'final로 변수를 선언할 때는 선언과 초기화를 동시에 해야 하며, 그렇지 않은 경우 컴파일 에러가 발생한다'고 알고 있었는데 제 코드가 왜 잘 동작하는지 이상해졌습니다🤔
자바의 정석에서 final 관련 내용을 찾아보니 , 인스턴스변수의 경우에는 예외적으로 생성자에서 초기화가 가능하다는 것을 배웠습니다. final인스턴스변수라도 각 인스턴스마다 다른 값을 가질 수 있어야 하기 때문이었네요.
질문을 던져 주시지 않았다면 깊게 공부하지 못했을 텐데, 코멘트 주셔서 감사합니다! 🙇‍♀️🙇‍♀️

honux77 replied:
JVM이 생성자를 만나야 객체의 인스턴스를 힙에 생성하기 때문에 final 인스턴스 변수는 생성자에서 초기화할 수 있습니다.
static final 변수는 그렇지 않겠죠?

bibi6666667 replied:
아 원리적으로도 그렇군요. JVM에 대해 더 자세히 공부해 보겠습니다. 감사합니다😀🙇‍♀️

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.ArrayList;
+
+import pieces.Pawn;
+
+public class Board {
+    private ArrayList<Pawn> pawns = new ArrayList<>();
```
List<Pawn> pawns = ... 으로 많이 사용합니다.

bibi6666667 replied:
그렇군요! 해당 부분 수정한 다음, 왜 ArrayList는 List로 업캐스팅(맞나요?)해서 사용하는지 구글링하고 학습했습니다. 
사실 아직도 완전히 이해되진 않아서, 우선 더 공부해 본 뒤에 목요일 수업 sli.do에 질문 남겨 보겠습니다!
코멘트 감사합니다🙇‍♀️🙇‍♀️

honux77 commented on src/test/java/chess/BoardTest.java:
```diff
@@ -0,0 +1,42 @@
+
+import org.junit.jupiter.api.*;
+
+import static org.junit.jupiter.api.Assertions.*;
```
assertEquall은 제거하고 asssertThat을 사용해 주세요.

bibi6666667 replied:
BoardTest와 PawnTest에서 `assertEqual()`을 사용했던 부분을 모두 `assertThat().isEqualTo()`로 변경했습니다.
아울러 AssertJ 에서 제공하는 여러 메서드들도 공부하고 정리하겠습니다.
코멘트 감사합니다!🙇‍♀️

honux77 commented on src/main/java/bibi/Pawn.java:
```diff
@@ -1,13 +1,18 @@
 
-public final class Pawn {
-    private String color;
+public class Pawn {
+    private final String color;
 
     Pawn(String color) {
         this.color = color;
```
지금 Pawn이 두 개인것 같네요. 불필요한 클래스는 제거해 주세요.

bibi6666667 replied:
앗.. 급한 마음에 제대로 확인하지 못하고 PR드려서 죄송합니다😭 수정하겠습니다!

honux77 commented on src/main/java/pieces/Pawn.java:
```diff
@@ -0,0 +1,19 @@
+    }
+
+    public Pawn() {
+        this.color = WHITE;
```
this(WHITE);

bibi6666667 replied:
this()를 처음 접했어서 구글링과 자바의 정석 통해 공부한 뒤 해당 부분 변경했습니다!
그런데 한 가지 질문이 있는데,
이렇게 `this.color = WHITE` 에서 `this(WHITE)`로 수정한 것은 기능 수정이 아닌 리팩토링에 해당하는 것이 맞나요?
코드 외형만 바뀌었을 뿐 결국 `Pawn(String color)` 생성자를 호출하기 떄문에 리팩토링이다. 라고 이해했는데,
가능하시다면 조언 부탁드립니다.
코멘트 주셔서 감사합니다 호눅스! 🙇‍♀️🙇‍♀️

honux77 replied:
넵 맞아요.

bibi6666667 replied:
그렇군요! 감사합니다😀

honux77 commented on src/test/java/chess/BoardTest.java:
```diff
@@ -0,0 +1,42 @@
+    @Test
+    @DisplayName("체스판에 흰색 폰을 추가 후 갯수와 목록을 확인한다.")
+    void putWhitePawn() {
+        Pawn white = new Pawn(Pawn.WHITE);
```
만들었던 기본 생성자를 활용해 보는 건 어때요?

bibi6666667 replied:
수정하겠습니다! 코드 중복만 없애는 게 아닌, 만들어 둔 모듈을 활용해 코드를 간결하게 하는 것도 리팩토링에 중요함을 기억헤 두겠습니다.
코멘트 감사합니다!🙇‍♀️

honux77 commented on src/test/java/pieces/PawnTest.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @Test
+    @DisplayName("색이 없는 폰 생성")
```
이름과 내용이 맞지 않아요.

bibi6666667 replied:
이 부분도 제대로 확인하지 않고 PR드려서 죄송합니다🙇‍♀️ 수정하겠습니다!

honux77 reviewed (changes requested):
수고하셨어요. 코드 개선해서 다시 리뷰 요청 해 주세요.

bibi6666667 committed:
[step2] createBlackPawn() 코드 수정

검은색 폰을 생성해야 하는데, 흰색 폰을 생성하고 있어
코드를 수정했습니다.

bibi6666667 committed:
[step2] 삭제) bibi패키지와 중복 클래스 Pawn 삭제

bibi6666667 committed:
[step2] 수정) pawns 선언을 List<Pawn> pawns로 변경

ArrayList를 선언해 사용하더라도 List pawns = new ArrayList 와 같은
방식으로 선언하는 것이 유연성 면에서 더 좋다고 한다.

bibi6666667 committed:
[step2]수정)모든 assertEquals()를 assertThat()으로

테스트 클래스에 있는 모든 assertEquals() 메서드를
assertThat().isEqualTo()로 변경했습니다.

bibi6666667 committed:
[step2] 수정) Pawn의 기본 생성자 코드를 this(WHITE)로변경

this는 알았지만 this()는 피드백을 통해 처음 알게 되었습니다.
구글링, 자바의 정석을 통해 this()의 사용법을 배운 다음
Pawn 기본 생성자의 코드를 피드백 주신 대로 수정했습니다.

bibi6666667 committed:
[step2] 수정) putWhitePawn()이 사용하는 생성자 변경

Pawn기본 생성자가 흰색 폰을 만들기 떄문에,
putWhitePawn()에서 Pawn(String color)대신
Pawn()으로 객체를 생성하게 함으로써
코드를 간결하게 했습니다.

bibi6666667 commented:
@honux77 안녕하세요, 백엔드반 Bibi입니다.
미션2 PR에 피드백 주신 부분에 대해 공부하고, 코드 개선하여 다시 Pull 요청 드립니다.

<추가로 학습한 부분>
- `List 리스트 = ArrayList`로 업캐스팅? 하여 선언하는 이유
- `this()` 에 대해, 그리고 `this`와의 차이점
- AssertJ에서 제공하는 다양한 assert메소드들 및 사용법 (더 학습 필요)
- `final` 인스턴스 변수 관련 심화 학습

<느낀 점>
- 중복 코드만 피한다고 리팩토링이 다가 아니다. 만들어 둔 모듈을 활용해 코드를 간결하게 하자.
- 내 코드를 잘 살핀 다음 PR을 신중하게 해서 잔실수들을 예방하자..😅

저는 자바 관련 서적 진도 나가고 수업 듣는 것도 재밌지만, 이렇게 직접 코딩하고 피드백 받으며 찾아서 공부하니 지식이 와닿는 느낌이라 요즘 재미있게 공부하고 있습니다! 
그리고 깃 공부가 자동으로 되는 것도 참 좋은 학습방식이라고 느낍니다.👍

신경 써서 자세한 피드백과 코멘트 남겨 주시고, 좋은 질문도 던져 주셔서 감사합니다.🙇‍♀️🙇‍♀️
끝으로 즐거운 오후 되세요! 😀

honux77 reviewed (approved):
💯
